Coding standard

RyanBerge opened the pull request:
Update coding standard to reflect current thinking and add in ros-specific clauses.

RyanBerge committed:
Initial Coding Standard Proposal

RyanBerge committed:
Update coding standard based on feedback

RyanBerge committed:
CodingStandard: Add formatting standards

- Also link to ROS and Google style guides

collinsc commented on CodingStandard.md:
```diff
@@ -1,34 +1,120 @@
+##Formatting
+
+- All C++ and Python code should use 4 spaces instead of tabs.
+- C++ braces should always be placed on their own line.
```
Hey @RyanBerge @AdamuKaapan there is an edge case that I want thrown in here. Uniform braced initialization. I use it, it's use is encouraged by scott meyers, herb sutter, the ISO c++ standard guidelines, and bjarne strostrup. 

It avoids narrowing and widening errors, and eliminates the c++ most vexing parse.

E.G:  
Server server{ n, "do_action", boost::bind(&execute, _1, &server), false };

It would be silly to put on multiple lines. I want a clause allowing for uniform braced initialization.

AdamuKaapan replied:
I support this. I had to look up what it was first, I've never used those before, but that's a perfectly reasonable clause to add it seems like.

RyanBerge replied:
Sure. I don't think anyone intended to restrict brace-initialization to multiple lines.

collinsc reviewed (changes requested):
We need to account for uniform braced initialization syntax in the way word the brace clause.

AdamuKaapan commented on CodingStandard.md:
````diff
@@ -1,34 +1,120 @@
+```
+  - Pre-Processor Macros: `UPPER_CASE`
+  - Class, Struct, Union, and Enum Types: `CamelCase`
+  - Functions and Class Methods: `lower_case`
````
I'm curious for people's opinions on this: I've always loved `camelCase` for methods personally. The ROS guide says `camelCase`, the Google one says `CamelCase`, and we currently have `lower_case`. I was just curious to see what everyone else thought of this and if we wanted to change it possibly.

RyanBerge replied:
I don't really care.  I learned on the C# standard which is `camelCase` for private methods and functions and `CamelCase` for public methods, but at UASC we used `lower_case` for both methods and functions.  The only reason I can think of going either way is that `CamelCase` looks like a Type and `lower_case` looks like a variable, but a function is neither.

AdamuKaapan replied:
If we want to make it all look unique (going off of your last sentence), we'd have `CamelCase` for types, `lower_case` for variables, and `camelCase` for methods, that way they're all unique. I also think `lower_case` for private methods and `camelCase` for public ones would be alright, that seems reasonable to me.

collinsc replied:
Also CamelCase has been called PascalCase wherever I have seen it or upper camel case.
While the camelCase refers to this.

So to clarify from what I have read we are going to change this to:
Types 
+  - Pre-Processor Macros: `UPPER_CASE`
+  - Class, Struct, Union, and Enum Types: `PascalCase`
+  - Class Methods: `camelCase`
+ -  local and instance variables: `under_scored`
+ - Global variables:  `UPPER_CASE`

yes?

AdamuKaapan replied:
Sounds right, yeah.

RyanBerge replied:
Can we do `gCamelCase` for global variables to differentiate them for macros (we currently have no  rule for global variables, but it matches our `kCamelCase` rule for constants)?

collinsc replied:
We haven't used the Hungarian naming convention until that suggestion. I don't think it provides good consistency. Can we let context infer on that issue?

AdamuKaapan commented on CodingStandard.md:
````diff
@@ -1,34 +1,120 @@
-If the same question is likely to arise again later in the project, notify
-the Lead Software Engineer, so he or she can update this page for everyone's
-benefit.
+```
````
Maybe add a note saying there's an example here, without just jumping straight into code?

AdamuKaapan reviewed (commented):
Added a question about method naming conventions, also agree with @collinsc 's comment on uniform braced initialization.

collinsc reviewed (changes requested):
added comments requesting changes to be implemented, and asking for clarification on case.

collinsc commented:
@AdamuKaapan It's best to explicitly request changes when you have been assigned to review something than, just not explicitly approve or deny. It actually notifies the assignee.

AdamuKaapan commented:
@collinsc yeah, that's true. I guess the reason I didn't request changes here was because it was more just looking for an opinion, but I guess that still works here.

RyanBerge committed:
CodingStandard: Tweak coding standard

- Add exception for braced-initialization lists
- Also change naming convention for functions from lower_case to lowerCamelCase

RyanBerge commented:
@collinsc @AdamuKaapan  I've updated the pull request based on the feedback.

collinsc commented on CodingStandard.md:
```diff
@@ -1,34 +1,118 @@
+  - Functions and Class Methods: `lowerCamelCase`
+  - Compile-time constants (const, constexpr, enumerators, etc.): `kCamelCase`
+  - Variables and Function Parameters: `lower_case`
+  - Const member variables: `kCamelCase`
```
No, I am not a fan of hungarian naming convention we need to get that out of the coding standard.

It hurts readability, it obfuscates intent, it makes code harder to maintain, we need that 100% _out_ of our code base.

collinsc replied:
https://en.wikipedia.org/wiki/Hungarian_notation

collinsc replied:
It makes sense for large mature code bases, but not for this, it sounds like a recipie for a maintainability disaster.

collinsc reviewed (changes requested):
No way I will ever ever ever sign off on the hungarian naming convention. I thought I left a comment on this yesteday, but it looks like it didn't go through.

RyanBerge committed:
CodingStandard: Change naming for constants and functions
